src/deploy.sh

```diff
@@ -1,158 +1,331 @@
-    echo
-    echo
-    exit 1
-  fi
-  
-  sleep 10 # Wait until deployment is created
-  DEPLOY_TIMEOUT_PERIOD=$((DEPLOY_TIMEOUT_PERIOD + 10))
-done
+# ----- Create app spec file -----
+h1 "Step 4: Creating App Spec for CodeDeploy"
+runCommand "envsubst < app-spec.tpl.json > app-spec.json" \
+            "Create app-spec file failed" \
+            "Create app-spec file"
 
-echo "---> For more info: https://$AWS_DEFAULT_REGION.console.aws.amazon.com/codesuite/codedeploy/deployments/$DEPLOYMENT_ID"
+isjsonValid "app-spec.json"
+info "App spec file:"
+cat app-spec.json | jq
 
-/work/tail-ecs-events.py &
-TAIL_ECS_EVENTS_PID=$!
+# ----- Create Deployment -----
+h1 "Step 5: Creating Deployment"
+APPLICATION_NAME=$CLUSTER_NAME-$APP_NAME
+DEPLOYMENT_CONFIG_NAME=CodeDeployDefault.ECSAllAtOnce
+DEPLOYMENT_GROUP=$CLUSTER_NAME-$APP_NAME
 
-RET=0
+# TODO: Check if is there any deployment in progress
 
-while [ "$(aws deploy get-deployment --deployment-id $DEPLOYMENT_ID --query deploymentInfo.status --output text)" == "Created" ]
-do
-  sleep 1
-done
+DEPLOYMENT_CMD="aws deploy create-deployment \
+    --output json \
+    --application-name $APPLICATION_NAME \
+    --deployment-config-name $DEPLOYMENT_CONFIG_NAME \
+    --deployment-group-name $DEPLOYMENT_GROUP \
+    --description Deployment \
+    --revision file://app-spec.json"
 
-echo "---> Deployment created!"
+DEPLOYMENT_OUTPUT=""
+runCommand "$DEPLOYMENT_CMD" \
+           "Deployment of application \"$APPLICATION_NAME\" on deployment group \"$DEPLOYMENT_GROUP\" failed" \
+           "" \
+           DEPLOYMENT_OUTPUT
 
+DEPLOYMENT_ID=$(echo $DEPLOYMENT_OUTPUT | jsonValue 'deploymentId' | tr -d ' ')
+success "Successfully created deployment: \"$DEPLOYMENT_ID\""
+note "For more info, you can follow your deployment at: https://$AWS_DEFAULT_REGION.console.aws.amazon.com/codesuite/codedeploy/deployments/$DEPLOYMENT_ID"
+
+
+# ----- Monitor Deployment -----
+h1 "Step 6: Deployment Overview"
+
+DEPLOY_TIMEOUT=
 DEPLOY_TIMEOUT_PERIOD=0
-DEPLOY_TIMEOUT_REACHED=0
-
-while [ "$(aws deploy get-deployment --deployment-id $DEPLOYMENT_ID --query deploymentInfo.status --output text)" == "InProgress" ]
-do
-  if [ "$DEPLOY_TIMEOUT_PERIOD" -ge "${DEPLOY_TIMEOUT:-900}" ]; then
-    echo "---> WARNING: Timeout reached. Rolling back deployment..."
-    aws deploy stop-deployment --deployment-id $DEPLOYMENT_ID --auto-rollback-enabled
-    DEPLOY_TIMEOUT_REACHED=1
-  fi
-  sleep 1
-  DEPLOY_TIMEOUT_PERIOD=$((DEPLOY_TIMEOUT_PERIOD + 1))
+
+DEPLOYMENT_GET="aws deploy get-deployment --output json --deployment-id \"$DEPLOYMENT_ID\""
+h2 "Monitoring deployment \"$DEPLOYMENT_ID\" for \"$APPLICATION_NAME\" on deployment group $DEPLOYMENT_GROUP ..."
+info "$DEPLOYMENT_GET"
+
+TASK_SET_ID=""
+
+while [ "${TASK_SET_ID}" == "" ]; do
+  TASK_SET_ID=$(aws ecs describe-services --cluster $CLUSTER_NAME --service $APP_NAME --query "services[0].taskSets[?status == 'ACTIVE'].id" --output text)
```

Comment:
i believe we should have a sleep here to avoid flooding the aws api

Reply:
Agree 👍